services/catalog/src/simcore_service_catalog/services/access_rights.py

```diff
@@ -0,0 +1,167 @@
+
+    # try to find the owner
+    possible_owner_email = [service.contact] + [
+        author.email for author in service.authors
+    ]
+
+    for user_email in possible_owner_email:
+        possible_gid = await groups_repo.get_user_gid_from_email(user_email)
+        if possible_gid:
+            if not owner_gid:
+                owner_gid = possible_gid
+    if not owner_gid:
+        logger.warning("service %s:%s has no owner", service.key, service.version)
+    else:
+        reader_gids.append(owner_gid)
+
+    # we add the owner with full rights, unless it's everyone
+    default_access_rights = [
+        ServiceAccessRightsAtDB(
+            key=service.key,
+            version=service.version,
+            gid=gid,
+            execute_access=True,
+            write_access=(gid == owner_gid),
+            product_name=app.state.settings.access_rights_default_product_name,
+        )
+        for gid in set(reader_gids)
+    ]
+
+    # Patch releases inherit access rights from previous version
+
+    return (owner_gid, default_access_rights)
+
+
+async def evaluate_auto_upgrade_policy(
+    service_metadata: ServiceDockerData, services_repo: ServicesRepository
+):
+    # AUTO-UPGRADE PATCH policy:
+    #
+    #  - Any new patch released, inherits the access rights from previous compatible version
+    #  - TODO: add as option in the publication contract, i.e. in ServiceDockerData
+    #  - Does NOT apply to front-end services
+    #
+    # SEE https://github.com/ITISFoundation/osparc-simcore/issues/2244)
+    #
+    if _is_frontend_service(service_metadata):
+        return []
+
+    service_access_rights = []
+    new_version: Version = as_version(service_metadata.version)
+    latest_releases = await services_repo.list_service_releases(
+        service_metadata.key,
+        major=new_version.major,
+        minor=new_version.minor,
+        limit_count=1,
+    )
+    assert len(latest_releases) <= 1  # nosec
+
+    if latest_releases:
+        latest_patch = latest_releases[0]
+
+        if is_patch_release(new_version, latest_patch.version):
+            previous_access_rights = await services_repo.get_service_access_rights(
+                latest_patch.key, latest_patch.version
+            )
+
+            for access in previous_access_rights:
+                service_access_rights.append(
+                    access.copy(
+                        exclude={"created", "modified"},
+                        update={"version": service_metadata.version},
+                        deep=True,
+                    )
+                )
+
+    return service_access_rights
+
+
+def merge_access_rights(
+    access_rights: List[ServiceAccessRightsAtDB],
+) -> List[ServiceAccessRightsAtDB]:
+    # TODO: not generic enought.
+    # TODO: probably a lot of room to optimize
+    merged = {}
+    for access in access_rights:
+        resource = access.get_resource()
+        flags = merged.get(resource)
```

Comment:
ok I see now where you use this. but I still find this naming somewhat confusing.

Reply:
I tend to agree with you ... i refactored it. Those functions are just helpers to do the reduction operations on access rights items. They are not general enough to be part of the model.